zuul-core/src/main/java/com/netflix/zuul/netty/server/http2/Http2ContentLengthEnforcingHandler.java

```diff
@@ -52,39 +53,51 @@ public final class Http2ContentLengthEnforcingHandler extends ChannelInboundHand
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         if (msg instanceof HttpRequest) {
-            HttpRequest req = (HttpRequest) msg;
-            List<String> lengthHeaders = req.headers().getAll(HttpHeaderNames.CONTENT_LENGTH);
-            if (lengthHeaders.size() > 1) {
-                ctx.writeAndFlush(new DefaultHttp2ResetFrame(Http2Error.PROTOCOL_ERROR));
-                return;
-            } else if (lengthHeaders.size() == 1) {
-                expectedContentLength = Long.parseLong(lengthHeaders.get(0));
-                if (expectedContentLength < 0) {
+            try {
+                HttpRequest req = (HttpRequest) msg;
+                List<String> lengthHeaders = req.headers().getAll(HttpHeaderNames.CONTENT_LENGTH);
+                if (lengthHeaders.size() > 1) {
+                    ctx.writeAndFlush(new DefaultHttp2ResetFrame(Http2Error.PROTOCOL_ERROR));
+                    return;
+                } else if (lengthHeaders.size() == 1) {
+                    expectedContentLength = Long.parseLong(lengthHeaders.get(0));
+                    if (expectedContentLength < 0) {
+                        // TODO(carl-mastrangelo): this is not right, but meh.  Fix this to return a proper 400.
+                        ctx.writeAndFlush(new DefaultHttp2ResetFrame(Http2Error.PROTOCOL_ERROR));
+                        return;
+                    }
+                }
+                if (hasContentLength() && HttpUtil.isTransferEncodingChunked(req)) {
                     // TODO(carl-mastrangelo): this is not right, but meh.  Fix this to return a proper 400.
                     ctx.writeAndFlush(new DefaultHttp2ResetFrame(Http2Error.PROTOCOL_ERROR));
                     return;
                 }
-            }
-            if (hasContentLength() && HttpUtil.isTransferEncodingChunked(req)) {
-                // TODO(carl-mastrangelo): this is not right, but meh.  Fix this to return a proper 400.
-                ctx.writeAndFlush(new DefaultHttp2ResetFrame(Http2Error.PROTOCOL_ERROR));
-                return;
+            } finally {
+                ReferenceCountUtil.safeRelease(msg);
```

Comment:
Since we are firing a channel read on Line 103 - I don't think we want to release in those cases, rather just when hitting the writeAndFlush/return blocks

Reply:
Ah yeah, I was trying to minimize the diff but probably need to be inside each if block